## Cytoscape WorkFlows

Collection of workflows as R Notebooks that use functionaility in R to process data and cyRest to communicate directly with Cytoscape.

# Enrichment Map Pipeline
Set of R notebooks to transform expression data to a ranked list and run them through Pathay enrichment pipeline. Pathway enrichment analysis helps gain mechanistic insight into large gene lists typically resulting from genome scale (â€“omics) experiments. It identifies biological pathways that are enriched in the gene list more than expected by chance. We explain pathway enrichment analysis and present a practical step-by-step guide to help interpret gene lists. The protocol is designed for biologists with no prior bioinformatics training and uses freely available software including g:Profiler, GSEA, Cytoscape and Enrichment Map.

## Main Pipeline Steps
 1. [Protocol 2 - Create Enrichment Map](https://github.com/BaderLab/Cytoscape_workflows/EnrichmentMapPipeline/Protocol2_createEM.html)

# Cell Cell Interactions Workflow

Using a set of proteins designated as receptors, and ligands defined with a set of GO terms calculate the set of interactions that represent cell-cell interactions (for example Ligand-receptor, receptor-receptor, ...). This analysis is not limited to Cell-Cell interactions. You can define your own protein types, either manually or by choosing different go terms, and create your customized protein-protein interaction network.

[editor on GitHub](https://github.com/BaderLab/Cytoscape_workflows/edit/master/README.md)
