---
title: "Supplementary Protocol 1 – create a gene list by analyzing gene expression data from RNA-seq with edgeR"
author: "Ruth Isserlin"
date: "`r format(Sys.Date())`"
    output:
      html_document:
        keep_md: true
---

This part of the supplementary protocol demonstrates filtering and scoring RNA-Seq data using normalized RNA-Seq count data with the edgeR R package. The protocol can be used to produce input data for pathway enrichment methods like g:Profiler, GSEA and others. This RNA-Seq analysis protocol follows conceptually similar steps to microarray analysis shown above. 

1. Load required Bioconductor packages into R. 
```{r}
tryCatch(expr = { library("edgeR")}, 
         error = function(e) { 
           source("https://bioconductor.org/biocLite.R")
           biocLite("edgeR")}, 
         finally = library("edgeR"))

working_dir <- paste(getwd(),"data",sep="/")

#The field in the class definition file that defines the classes of the data.
data_classes <- "SUBTYPE"
```


2. Load the expression dataset of 216 tumours, with 107 classified as Immunoreactive and 109 classified as Mesenchymal samples.

```{r}
RNASeq <- read.table(  paste(working_dir,"Supplementary_Table12_TCGA_RNASeq_rawcounts.txt",sep="/"),  header = TRUE, sep = "\t", quote="\"", stringsAsFactors = FALSE)
```


3. Load subtype classification of samples. To calculate differential expression, we need to define at least 2 classes of samples. A common experimental design involves cases and controls but any two classes can be used. The current dataset is divided into mesenchymal and immunoreactive classes (Supplementary Table 4, third column). After loading the matrix, check that the column names of the expression matrix and class definitions are equal. 

```{r}
classDefinitions_RNASeq <- read.table( paste(working_dir, "Supplementary_Table13_RNASeq_classdefinitions.txt", sep="/"), header = TRUE, sep = "\t", quote="\"", stringsAsFactors = FALSE)
```

4. Filter RNA-seq reads. RNA-Seq data are processed following the edgeR protocol15 that filters reads based on the counts per million (CPM) statistic. RNA-Seq read counts are converted to CPM values and genes with CPM > 1 in at least 42 of the samples are retained for further study (42 is the minimal sample size in the classes).  This step removes genes with very low read counts that are likely not expressed in the majority of samples and cause noise in the data. Note CPM conversion is used to remove low counts while differential gene expression analysis is based on initial normalized read counts. 
```{r}
cpms <- cpm(RNASeq)
keep <- rowSums(cpms > 1) >= 50
counts <- RNASeq [keep,]
```

5. Exclude genes with missing symbols. Gene entries in the dataset containing ‘?’ or starting with LOC are excluded as they represent non-annotated genes or other loci that are not present in pathway databases.
```{r}
exclude <- grep("\\?|^LOC", rownames(counts), value=T)
counts <- counts[which(!rownames(counts) %in% exclude),]
```


6. Data normalization, dispersion analysis and differential expression analysis is performed with a simple design as described in the edgeR protocol15.
```{r}
d <- DGEList(counts=counts, group=classDefinitions_RNASeq[,data_classes])
d <- calcNormFactors(d)
mds_output <- plotMDS(d, labels=classDefinitions_RNASeq[,data_classes], 
col= c("darkgreen","blue","red","orange")[factor(classDefinitions_RNASeq[,data_classes])])
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)
de <- exactTest(d, pair=c("Mesenchymal","Immunoreactive"))
tt <- topTags(de,n=nrow(d))


```
```{r}
classes <- factor(classDefinitions_RNASeq[,data_classes])
modelDesign <- model.matrix(~ 0 + classes)

contrast_mesenvsimmuno <- makeContrasts(mesenvsimmuno ="classesMesenchymal-classesImmunoreactive",levels=modelDesign)
fit_glm <- glmFit(d,modelDesign)
mesenvsimmuno <- glmLRT(fit_glm , contrast = contrast_mesenvsimmuno)
tt_mesenvsimmuno <- topTags(mesenvsimmuno,n=nrow(d))

```

```{r}
classes <- factor(classDefinitions_RNASeq[,data_classes])
modelDesign <- model.matrix(~ 0 + classes)

contrast_immuno <- makeContrasts(immunovsrest ="classesImmunoreactive-(classesMesenchymal + classesProliferative +classesDifferentiated)/3",levels=modelDesign)
fit_glm <- glmFit(d,modelDesign)
immunovsrest <- glmLRT(fit_glm , contrast = contrast_immuno)
tt_immunovsrest <- topTags(immunovsrest,n=nrow(d))

contrast_mesen <- makeContrasts( mesenvsrest = "classesMesenchymal-(classesImmunoreactive + classesProliferative +classesDifferentiated)/3",levels=modelDesign)
fit_glm <- glmFit(d,modelDesign)
mesenvsrest <- glmLRT(fit_glm , contrast = contrast_mesen)
tt_mesenvsrest <- topTags(mesenvsrest,n=nrow(d))

contrast_prolif <- makeContrasts( prolifvsrest = "classesProliferative-(classesMesenchymal + classesImmunoreactive +classesDifferentiated)/3",levels=modelDesign)
fit_glm <- glmFit(d,modelDesign)
prolifvsrest <- glmLRT(fit_glm , contrast = contrast_prolif)
tt_prolifvsrest <- topTags(prolifvsrest,n=nrow(d))

contrast_diff <- makeContrasts( diffvsrest = "classesDifferentiated-(classesMesenchymal + classesImmunoreactive +classesProliferative)/3",levels=modelDesign)
fit_glm <- glmFit(d,modelDesign)
diffvsrest <- glmLRT(fit_glm , contrast = contrast_diff)
tt_diffvsrest <- topTags(diffvsrest,n=nrow(d))
```


7a. Create the gene list for use in g:Profiler or another thresholded enrichment tool. The list may comprise all genes that have a significant FDR-corrected p-value (code shown below), all significant and FDR-corrected up-regulated genes and all down-regulated genes separately, or some other combination of thresholds. Also see analogous step in the microarray protocol.
```{r}
tt <- tt_mesenvsimmuno
select_genes = which(tt$table$FDR<0.05)
length(select_genes)
topgenes_qvalue005 <- unlist(lapply( rownames(tt$table)[select_genes], 
function(data) {unlist(strsplit(data,"\\|"))[1]}))
head(topgenes_qvalue005)
write.table(topgenes_qvalue005, 
paste(working_dir,"MesenvsImmuno_RNAseq_allsignificantgenes.txt",sep="/"), 
col.names=FALSE, sep="\t", row.names=FALSE, quote=FALSE)
```

7b. Create a two-column rank (.RNK) file of all gene IDs and corresponding scores to for GSEA pre-ranked analysis. One option is to rank genes by t-statistic of differential gene expression. GSEA will look for enrichment in the set of most differentially expressed genes at the top of the list as well as those at the bottom of the list. Genes at the top of the list are more highly expressed in class A of samples (e.g., mesenchymal) while genes at the bottom are highly expressed in class B (e.g., immunoreactive). An alternative score can be computed by multiplying direction (sign) of fold change and logarithm of p-value for each gene.
```{r}
ranks_RNAseq = sign(tt$table$logFC) * -log10(tt$table$PValue)
genenames <- unlist(lapply( rownames(tt$table), function(data) 
  {unlist(strsplit(data,"\\|"))[1]}))
geneids <- unlist(lapply( rownames(tt$table),  function(data) {unlist(strsplit(data,"\\|"))[2]})) 
ranks_RNAseq <- cbind(genenames, ranks_RNAseq)
colnames(ranks_RNAseq) <- c("GeneName","rank")
write.table(ranks_RNAseq, paste(working_dir,"Supplementary_Table2_MesenvsImmuno_RNASeq_ranks.rnk",sep="/"), 
col.name = TRUE, sep="\t", row.names = FALSE, quote = FALSE)
```




8a. Create an expression file for the enrichment map and save it to a file in the working folder. The optional expression file is similar to the expression matrix except for an additional column on the left edge of the matrix. The field often includes gene description however any text value can be added.
```{r}
normalized_expression_RNAseq <- cpm(d, normalized.lib.size=TRUE)

#From the rownames parse out the gene name and the geneids
genenames <- unlist(lapply( rownames(normalized_expression_RNAseq), 
function(data) {unlist(strsplit(data,"\\|"))[1]}))
geneids <- unlist(lapply( rownames(normalized_expression_RNAseq), 
function(data) {unlist(strsplit(data,"\\|"))[2]}))

#create a matrix with the gene names and gene ids
include_genes <- grep("\\?|^LOC", genenames, invert=T)

EM_expressionFile_RNAseq <- data.frame(Name = genenames[include_genes ], normalized_expression_RNAseq[include_genes ,])
rownames(EM_expressionFile_RNAseq) <- rownames(normalized_expression_RNAseq[include_genes ,])
colnames(EM_expressionFile_RNAseq) <- substring(colnames(EM_expressionFile_RNAseq),1,12)


#Add descriptions instead of geneids
tryCatch(expr = { library("biomaRt")}, 
         error = function(e) { 
           source("https://bioconductor.org/biocLite.R")
           biocLite("biomaRt")}, 
         finally = library("biomaRt"))
mart = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")

genes = getBM(attributes = c( 'hgnc_symbol', 'description'), filters='hgnc_symbol', 
              values=genenames, mart=mart);
genes$description = gsub("\\[Source.*", "", genes$description);

EM_expressionFile_RNAseq <- merge(genes,EM_expressionFile_RNAseq,  all.y=TRUE,by.x=1, by.y=1)
colnames(EM_expressionFile_RNAseq)[1] <- "Name"
colnames(EM_expressionFile_RNAseq)[2] <- "Description"

write.table(EM_expressionFile_RNAseq, paste(working_dir,"Supplementary_Table6_TCGA_OV_RNAseq_expression.txt", sep="/"),
            col.name=TRUE,sep="\t", row.names=FALSE, quote=FALSE)

#write out a GSEA classes file. (optional)
fileConn <- file(paste(working_dir,"Supplementary_Table9_TCGA_OV_RNAseq_classes.cls",sep="/"))
writeLines(c(paste(length(classDefinitions_RNASeq[,data_classes]), "4 1"), 
             paste("# ", unique(classDefinitions_RNASeq[,data_classes])[1], " ",
                   unique(classDefinitions_RNASeq[,data_classes])[2], " ",
                   unique(classDefinitions_RNASeq[,data_classes])[3], " ",
                   unique(classDefinitions_RNASeq[,data_classes])[4])), fileConn)
write.table(t(classDefinitions_RNASeq[,data_classes]), 
            paste(working_dir,"Supplementary_Table9_TCGA_OV_RNAseq_classes.cls",sep="/"), col.name=FALSE, sep="\t",
            row.names=FALSE, quote=FALSE, append=TRUE)
close(fileConn)
```


8b. Examine gene expression data using heat maps. Heat maps can easily show the separation between sample classes, labeled by colors in the heat map header. By limiting to the most significantly differentially expressed list of genes (FDR-corrected p<0.05) we can verify whether the scoring accurately separates class A from class B. 
```{r}
tryCatch(expr = { library("pheatmap")}, 
         error = function(e) { 
          install.packages("pheatmap")}, 
         finally = library("pheatmap"))
tryCatch(expr = { library("RColorBrewer")}, 
         error = function(e) { 
          install.packages("RColorBrewer")}, 
         finally = library("RColorBrewer"))

annotation_col <- data.frame(SUBTYPE=factor(classDefinitions_RNASeq[,data_classes]))
rownames(annotation_col) <- classDefinitions_RNASeq[,2]

ann_colors = list(SUBTYPE = c(Immunoreactive="#1B9E77", Mesenchymal="#D95F02",Proliferative = "yellow",Differentiated="orange"))
col.pal <- rev(brewer.pal(11, "RdBu"))

genes_to_select <- unlist(lapply( rownames(tt$table)[which(tt$table$FDR<0.05)], function(data) 
  {unlist(strsplit(data,"\\|"))[1]}))

matrix_for_heatmap <- as.matrix(EM_expressionFile_RNAseq[which(EM_expressionFile_RNAseq[,1] %in% genes_to_select ),3:dim(EM_expressionFile_RNAseq)[2] ])

class(matrix_for_heatmap) <- "numeric"
matrix_for_heatmap[matrix_for_heatmap == 0] <- 0.0000001
pheatmap(matrix_for_heatmap, color=col.pal, scale="row", 
         kmeans_k=NA, show_rownames=FALSE, show_colnames=FALSE, 
         main="heatmap top genes(Mesen vs Immuno)", cluster_rows=TRUE, 
         cluster_cols=FALSE, clustering_distance_rows="correlation",
         annotation_col=annotation_col, annotation_colors=ann_colors)


```

Create the other files as well so all the comparisons can be compared.
```{r}
#Immuno vs rest
tt <- tt_immunovsrest
select_genes = which(tt$table$FDR<0.05)
length(select_genes)
topgenes_qvalue005 <- unlist(lapply( rownames(tt$table)[select_genes], 
function(data) {unlist(strsplit(data,"\\|"))[1]}))
head(topgenes_qvalue005)
write.table(topgenes_qvalue005, 
paste(working_dir,"ImmunovsRest_RNAseq_allsignificantgenes.txt",sep="/"), 
col.names=FALSE, sep="\t", row.names=FALSE, quote=FALSE)

ranks_RNAseq = sign(tt$table$logFC) * -log10(tt$table$PValue)
genenames <- unlist(lapply( rownames(tt$table), function(data) 
  {unlist(strsplit(data,"\\|"))[1]}))
geneids <- unlist(lapply( rownames(tt$table),  function(data) {unlist(strsplit(data,"\\|"))[2]})) 
ranks_RNAseq <- cbind(genenames, ranks_RNAseq)
colnames(ranks_RNAseq) <- c("GeneName","rank")
write.table(ranks_RNAseq, paste(working_dir,"ImmunovsRest_RNASeq_ranks.rnk",sep="/"), 
col.name = TRUE, sep="\t", row.names = FALSE, quote = FALSE)


#Mesen vs rest data
tt <- tt_mesenvsrest
select_genes = which(tt$table$FDR<0.05)
length(select_genes)
topgenes_qvalue005 <- unlist(lapply( rownames(tt$table)[select_genes], 
function(data) {unlist(strsplit(data,"\\|"))[1]}))
head(topgenes_qvalue005)
write.table(topgenes_qvalue005, 
paste(working_dir,"MesenvsRest_RNAseq_allsignificantgenes.txt",sep="/"), 
col.names=FALSE, sep="\t", row.names=FALSE, quote=FALSE)

ranks_RNAseq = sign(tt$table$logFC) * -log10(tt$table$PValue)
genenames <- unlist(lapply( rownames(tt$table), function(data) 
  {unlist(strsplit(data,"\\|"))[1]}))
geneids <- unlist(lapply( rownames(tt$table),  function(data) {unlist(strsplit(data,"\\|"))[2]})) 
ranks_RNAseq <- cbind(genenames, ranks_RNAseq)
colnames(ranks_RNAseq) <- c("GeneName","rank")
write.table(ranks_RNAseq, paste(working_dir,"MesenvsRest_RNASeq_ranks.rnk",sep="/"), 
col.name = TRUE, sep="\t", row.names = FALSE, quote = FALSE)

#Differentiated vs rest data
tt <- tt_diffvsrest
select_genes = which(tt$table$FDR<0.05)
length(select_genes)
topgenes_qvalue005 <- unlist(lapply( rownames(tt$table)[select_genes], 
function(data) {unlist(strsplit(data,"\\|"))[1]}))
head(topgenes_qvalue005)
write.table(topgenes_qvalue005, 
paste(working_dir,"DiffvsRest_RNAseq_allsignificantgenes.txt",sep="/"), 
col.names=FALSE, sep="\t", row.names=FALSE, quote=FALSE)

ranks_RNAseq = sign(tt$table$logFC) * -log10(tt$table$PValue)
genenames <- unlist(lapply( rownames(tt$table), function(data) 
  {unlist(strsplit(data,"\\|"))[1]}))
geneids <- unlist(lapply( rownames(tt$table),  function(data) {unlist(strsplit(data,"\\|"))[2]})) 
ranks_RNAseq <- cbind(genenames, ranks_RNAseq)
colnames(ranks_RNAseq) <- c("GeneName","rank")
write.table(ranks_RNAseq, paste(working_dir,"DiffvsRest_RNASeq_ranks.rnk",sep="/"), 
col.name = TRUE, sep="\t", row.names = FALSE, quote = FALSE)

#Proliferative vs rest data
tt <- tt_prolifvsrest
select_genes = which(tt$table$FDR<0.05)
length(select_genes)
topgenes_qvalue005 <- unlist(lapply( rownames(tt$table)[select_genes], 
function(data) {unlist(strsplit(data,"\\|"))[1]}))
head(topgenes_qvalue005)
write.table(topgenes_qvalue005, 
paste(working_dir,"ProlifvsRest_RNAseq_allsignificantgenes.txt",sep="/"), 
col.names=FALSE, sep="\t", row.names=FALSE, quote=FALSE)

ranks_RNAseq = sign(tt$table$logFC) * -log10(tt$table$PValue)
genenames <- unlist(lapply( rownames(tt$table), function(data) 
  {unlist(strsplit(data,"\\|"))[1]}))
geneids <- unlist(lapply( rownames(tt$table),  function(data) {unlist(strsplit(data,"\\|"))[2]})) 
ranks_RNAseq <- cbind(genenames, ranks_RNAseq)
colnames(ranks_RNAseq) <- c("GeneName","rank")
write.table(ranks_RNAseq, paste(working_dir,"ProlifvsRest_RNASeq_ranks.rnk",sep="/"), 
col.name = TRUE, sep="\t", row.names = FALSE, quote = FALSE)
```
