<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Ruth Isserlin" />

<meta name="date" content="2018-03-21" />

<title>Supplementary Protocol 1 – create a gene list by analyzing gene expression data from RNA-seq with edgeR</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #0000ff; } /* Keyword */
code > span.ch { color: #008080; } /* Char */
code > span.st { color: #008080; } /* String */
code > span.co { color: #008000; } /* Comment */
code > span.ot { color: #ff4000; } /* Other */
code > span.al { color: #ff0000; } /* Alert */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #008000; font-weight: bold; } /* Warning */
code > span.cn { } /* Constant */
code > span.sc { color: #008080; } /* SpecialChar */
code > span.vs { color: #008080; } /* VerbatimString */
code > span.ss { color: #008080; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #0000ff; } /* ControlFlow */
code > span.op { } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ff4000; } /* Preprocessor */
code > span.do { color: #008000; } /* Documentation */
code > span.an { color: #008000; } /* Annotation */
code > span.cv { color: #008000; } /* CommentVar */
code > span.at { } /* Attribute */
code > span.in { color: #008000; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 64px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 69px;
  margin-top: -69px;
}

.section h2 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h3 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h4 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h5 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h6 {
  padding-top: 69px;
  margin-top: -69px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Enrichment Map Protocol</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="https://baderlab.github.io/Cytoscape_workflows/">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="https://www.biorxiv.org/content/early/2017/12/12/232835">
    <span class="fa fa-newspaper-o"></span>
     
    bioRxiv paper
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-list"></span>
     
    Table of Contents
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Supplementary Protocols</li>
    <li>
      <a href="Download_TCGA_data.html">Download TCGA Data</a>
    </li>
    <li>
      <a href="supplemental_protocol1_rnaseq.html">Process RNASeq data</a>
    </li>
    <li>
      <a href="supplemental_protocol2_microarray.html">Process Microarray data</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Protocol 2</li>
    <li>
      <a href="Protocol2_createEM.html">Create Enrichment Map</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Supplementary Protocol 1 – create a gene list by analyzing gene expression data from RNA-seq with edgeR</h1>
<h4 class="author"><em>Ruth Isserlin</em></h4>
<h4 class="date"><em>2018-03-21</em></h4>

</div>


<p>This protocol processes RNA-seq data using the R programming environment and specialized packages from Bioconductor to create genes lists. The scripts are available for download and novice users can copy and paste commands into the R console. To create gene expression data for <strong>Protocol step 5B</strong>, we downloaded gene expression data from the Ovarian Serous Cystadenocarcinoma project of The Cancer Genome Atlas (TCGA)<span class="citation"><sup>1</sup></span>, <a href="http://cancergenome.nih.gov" class="uri">http://cancergenome.nih.gov</a> via the Genomic Data Commons (GDC) portal<span class="citation"><sup>2</sup></span> on 2017-06-14 using TCGABiolinks R package<span class="citation"><sup>3</sup></span>. The data includes 544 samples available as RMA-normalized microarray data (Affymetrix HG-U133A), and 309 samples available as RNA-seq data, with reads mapped to a reference genome using MapSplice<span class="citation"><sup>4</sup></span> and read counts per transcript determined using the RSEM method<span class="citation"><sup>5</sup></span>. RNA-seq data are labeled as ‘RNA-Seq V2’, see details at: <a href="https://wiki.nci.nih.gov/display/TCGA/RNASeq+Version+2" class="uri">https://wiki.nci.nih.gov/display/TCGA/RNASeq+Version+2</a>). The RNA-SeqV2 data consists of raw counts similar to regular RNA-seq but RSEM (RNA-Seq by Expectation Maximization) data can be used with the edgeR method.</p>
<div id="background-and-setup-for-all-supplmentary-protocols" class="section level1">
<h1><span class="header-section-number">1</span> Background and Setup for all Supplmentary Protocols</h1>
<div id="equipment" class="section level2">
<h2><span class="header-section-number">1.1</span> Equipment</h2>
<div id="hardware-requirements" class="section level3">
<h3><span class="header-section-number">1.1.1</span> Hardware requirements:</h3>
<ul>
<li>A recent personal computer with at least 8 gigabytes of memory (RAM). Software requirements:</li>
<li>The R statistical computing environment (<a href="http://www.r-project.org/" class="uri">http://www.r-project.org/</a>). We suggest using the integrated development environment RStudio (<a href="https://www.rstudio.com/" class="uri">https://www.rstudio.com/</a>).</li>
<li>Required R packages: Biobase, Limma and GSA are available from Bioconductor (<a href="https://www.bioconductor.org/" class="uri">https://www.bioconductor.org/</a>)</li>
</ul>
</div>
<div id="data-requirements" class="section level3">
<h3><span class="header-section-number">1.1.2</span> Data requirements:</h3>
<ul>
<li>Supplementary_Table10_TCGA_Microarray_rmanormalized.txt corresponds to the RMA normalized Affymetrix mRNA transcript expression data for serous ovarian cancer samples as downloaded from the GDC portal65. Normalization is required to compute differential gene expression values across subtypes, as performed in Supplementary Protocol 1A.</li>
<li>Supplementary_Table12_TCGA_RNASeq_rawcounts.txt corresponds to read counts per mRNA transcript determined using the RSEM method. These counts can be used to compare gene expression between subtypes using the edgeR analysis tool, as performed in Supplementary Protocol 1B. The counts are not pre-normalized and the normalization step using edgeR is part of the protocol.</li>
<li>Supplementary_Table11_Microarray_classdefinitions.txt and Supplementary_Table13_RNASeq_classdefinitions.txt define the subtype classification of ovarian cancer samples (immunoreactive, mesenchymal, differentiated, proliferative) (Verhaak et al.<span class="citation"><sup>6</sup></span> supplementary table 1, column 3). This information is used to extract two subgroups of interest, mesenchymal and immunoreactive.</li>
</ul>
</div>
<div id="equipment-setup" class="section level3">
<h3><span class="header-section-number">1.1.3</span> Equipment Setup</h3>
<ul>
<li>Download and install R from <a href="http://cran.r-project.org/" class="uri">http://cran.r-project.org/</a></li>
<li>Download and install RStudio from <a href="https://www.rstudio.com/" class="uri">https://www.rstudio.com/</a> (optional, but recommended)</li>
<li>Launch R or RStudio</li>
<li><p>Install required Bioconductor packages. Enter the following commands in the R command line (also see <a href="https://www.bioconductor.org/install/" class="uri">https://www.bioconductor.org/install/</a>):</p>
<ul>
<li>source(“<a href="http://www.Bioconductor.org/biocLite.R" class="uri">http://www.Bioconductor.org/biocLite.R</a>”)</li>
<li>biocLite(“BiocUpgrade”)</li>
<li>biocLite(c(“Biobase”,“limma”,”edgeR”,”GSA”,”locfit”))</li>
<li>install.packages(c(“pheatmap”, ”RColorBrewer”, ”gProfileR”, ”RJSONIO”, ”httr”))</li>
</ul></li>
<li><p>If the required packages are already installed you may receive a prompt to update these. The prompt window will ask you about updating the packages:</p>
<ul>
<li>Update all/some/none? [a/s/n]</li>
<li>Type ‘a’ without quotes and hit enter.</li>
</ul></li>
<li><p>Load libraries into the R session using the example below. Loading of libraries is required every time R is re-opened.</p>
<ul>
<li>library(“limma”)</li>
</ul></li>
<li>All R files are available at <a href="https://github.com/BaderLab/EM-tutorials-docker/tree/master/R_scripts" class="uri">https://github.com/BaderLab/EM-tutorials-docker/tree/master/R_scripts</a></li>
<li><p>Alternately, R notebooks (use R markdown and creates a notebook similar to Jupyter notebooks) of this protocol are available at <a href="https://github.com/BaderLab/Cytoscape_workflows/tree/master/EnrichmentMapPipeline" class="uri">https://github.com/BaderLab/Cytoscape_workflows/tree/master/EnrichmentMapPipeline</a></p></li>
</ul>
</div>
<div id="data-setup" class="section level3">
<h3><span class="header-section-number">1.1.4</span> Data setup</h3>
<ul>
<li>Download Supplementary Tables 10-13 to a dedicated folder of your computer. The first step of your R script will change the working directory of R to this folder.</li>
<li>As text editors sometimes add invisible characters to text copied from PDF files, copying and pasting from this document is not recommended. The R scripts or the R notebook available in the above URLs should be used instead.</li>
<li>Setting the current directory and loading packages (libraries) are the required first and second steps of each protocol. These are needed each time a new session is opened in R.</li>
</ul>
</div>
</div>
<div id="process-rnaseq-data" class="section level2">
<h2><span class="header-section-number">1.2</span> Process RNASeq data</h2>
<p>This part of the supplementary protocol demonstrates filtering and scoring RNA-Seq data using normalized RNA-Seq count data with the edgeR R package. The protocol can be used to produce input data for pathway enrichment methods like g:Profiler, GSEA and others. This RNA-Seq analysis protocol follows conceptually similar steps to microarray analysis shown above.</p>
<div id="load-required-packages" class="section level3">
<h3><span class="header-section-number">1.2.1</span> Load required packages</h3>
<ol style="list-style-type: decimal">
<li>Load required Bioconductor packages into R.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span>opts_knit<span class="op">$</span><span class="kw">set</span>(<span class="dt">cache =</span> <span class="ot">TRUE</span>)

<span class="kw">tryCatch</span>(<span class="dt">expr =</span> { <span class="kw">library</span>(<span class="st">&quot;edgeR&quot;</span>)}, 
         <span class="dt">error =</span> <span class="cf">function</span>(e) { 
           <span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
           <span class="kw">biocLite</span>(<span class="st">&quot;edgeR&quot;</span>)}, 
         <span class="dt">finally =</span> <span class="kw">library</span>(<span class="st">&quot;edgeR&quot;</span>))</code></pre></div>
<pre><code>## Loading required package: limma</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">working_dir &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">getwd</span>(),<span class="st">&quot;data&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>)

<span class="co">#The field in the class definition file that defines the classes of the data.</span>
data_classes &lt;-<span class="st"> &quot;SUBTYPE&quot;</span></code></pre></div>
</div>
<div id="load-expression-data" class="section level3">
<h3><span class="header-section-number">1.2.2</span> Load Expression Data</h3>
<ol start="2" style="list-style-type: decimal">
<li>Load the expression dataset of 300 tumours, with 79 classified as Immunoreactive, 72 classified as Mesenchymal, 69 classified as Differentiated, and 80 classified as Proliferative samples. The TCGA counts data was retrieved from the GDC<span class="citation"><sup>2</sup></span> and contained counts per mRNA transcript determined using the RSEM method for 19947 transcripts and 300 samples.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">RNASeq &lt;-<span class="st"> </span><span class="kw">read.table</span>( 
  <span class="kw">paste</span>(working_dir,<span class="st">&quot;Supplementary_Table12_TCGA_RNASeq_rawcounts.txt&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>),  
  <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;</span><span class="ch">\&quot;</span><span class="st">&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="load-subtype-information" class="section level3">
<h3><span class="header-section-number">1.2.3</span> Load subtype information</h3>
<ol start="3" style="list-style-type: decimal">
<li>Load subtype classification of samples. To calculate differential expression, we need to define at least two sample classes. A common experimental design involves cases and controls but any two classes can be used. The current dataset is divided into mesenchymal and immunoreactive classes (class definitions were obtained from Verhaak et al.<span class="citation"><sup>6</sup></span> Supplementary Table 1, third column). After loading the matrix, check that the column names of the expression matrix and class definitions are equal.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">classDefinitions_RNASeq &lt;-<span class="st"> </span><span class="kw">read.table</span>( 
  <span class="kw">paste</span>(working_dir, <span class="st">&quot;Supplementary_Table13_RNASeq_classdefinitions.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
  <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;</span><span class="ch">\&quot;</span><span class="st">&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="filter-data" class="section level3">
<h3><span class="header-section-number">1.2.4</span> Filter Data</h3>
<ol start="4" style="list-style-type: decimal">
<li>Filter RNA-seq reads. RNA-seq data are processed following the edgeR protocol15 that filters reads based on the counts per million (CPM) statistic. RNA-seq read counts are converted to CPM values and genes with CPM &gt; 1 in at least 50 of the samples are retained for further study (50 is the minimal sample size in the classes). This step removes genes with very low read counts that are likely not expressed in the majority of samples and cause noise in the data. Note, CPM filtering is used to remove low counts while differential gene expression analysis is based on normalized read counts which are generated below (step 6).</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cpms &lt;-<span class="st"> </span><span class="kw">cpm</span>(RNASeq)
keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(cpms <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>) <span class="op">&gt;=</span><span class="st"> </span><span class="dv">50</span>
counts &lt;-<span class="st"> </span>RNASeq[keep,]</code></pre></div>
</div>
<div id="normalization-and-dispersion" class="section level3">
<h3><span class="header-section-number">1.2.5</span> Normalization and Dispersion</h3>
<ol start="5" style="list-style-type: decimal">
<li>Data normalization, dispersion analysis is performed on the entire dataset.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># create data structure to hold counts and subtype information for each sample.</span>
d &lt;-<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">counts=</span>counts, <span class="dt">group=</span>classDefinitions_RNASeq<span class="op">$</span>SUBTYPE)

<span class="co">#Normalize the data</span>
d &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(d)

<span class="co">#create multidimensional scaling(MDS) plot.  The command below will automatically </span>
<span class="co"># generate the plot containing all samples where each subtype is a different color.  </span>
<span class="co">#Ideally there should be a good separation between the different classes.</span>
mds_output &lt;-<span class="st"> </span><span class="kw">plotMDS</span>(d, <span class="dt">labels=</span>classDefinitions_RNASeq<span class="op">$</span>SUBTYPE, 
<span class="dt">col=</span> <span class="kw">c</span>(<span class="st">&quot;darkgreen&quot;</span>,<span class="st">&quot;blue&quot;</span>,<span class="st">&quot;red&quot;</span>, <span class="st">&quot;orange&quot;</span>)[<span class="kw">factor</span>(classDefinitions_RNASeq<span class="op">$</span>SUBTYPE)])</code></pre></div>
<p><img src="supplemental_protocol1_rnaseq_files/figure-html/unnamed-chunk-6-1.png" /><!-- --></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#calculate dispersion</span>
d &lt;-<span class="st"> </span><span class="kw">estimateCommonDisp</span>(d)
d &lt;-<span class="st"> </span><span class="kw">estimateTagwiseDisp</span>(d)</code></pre></div>
</div>
<div id="filter-unannotated-genes" class="section level3">
<h3><span class="header-section-number">1.2.6</span> Filter unannotated genes</h3>
<ol start="6" style="list-style-type: decimal">
<li>(Optional) Exclude genes with missing symbols or uncharacterized genes. In this example gene entries in the dataset containing ‘?’ or starting with LOC are excluded as they represent non-annotated genes or other loci that are not present in pathway databases. The frequency of these and other non protein coding entries in your dataset will depend on the database used to align your RNASeq data.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#the below regular expression excludes gene names that are ? or that start with LOC</span>
<span class="co"># any number ofadditional terms can be added to the regular expresion, for example </span>
<span class="co"># to exclude any genes that start with &quot;His&quot; add |^His to the regular expression</span>
exclude &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">?|^LOC&quot;</span>, <span class="kw">rownames</span>(d), <span class="dt">value=</span>T)
d &lt;-<span class="st"> </span>d[<span class="kw">which</span>(<span class="op">!</span><span class="kw">rownames</span>(d) <span class="op">%in%</span><span class="st"> </span>exclude),]</code></pre></div>
</div>
<div id="calculate-differential-expression" class="section level3">
<h3><span class="header-section-number">1.2.7</span> Calculate Differential expression</h3>
<ol start="7" style="list-style-type: decimal">
<li>Differential expression analysis is performed with a simple design as described in the edgeR protocol<span class="citation"><sup>7</sup></span>.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#calculate differential expression statistics with a simple design</span>
de &lt;-<span class="st"> </span><span class="kw">exactTest</span>(d, <span class="dt">pair=</span><span class="kw">c</span>(<span class="st">&quot;Immunoreactive&quot;</span>,<span class="st">&quot;Mesenchymal&quot;</span>))
tt_exact_test &lt;-<span class="st"> </span><span class="kw">topTags</span>(de,<span class="dt">n=</span><span class="kw">nrow</span>(d))

<span class="co">#alternately you can also use the glm model using contrasts.  </span>
<span class="co">#For a simple 2 class comparison this is not required but if you want to compare</span>
<span class="co"># 1 class to the remaining 3 classes then this sort of model is useful.</span>
classes &lt;-<span class="st"> </span><span class="kw">factor</span>(classDefinitions_RNASeq[,data_classes])
modelDesign &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="dv">0</span> <span class="op">+</span><span class="st"> </span>classes)

contrast_mesenvsimmuno &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(
                  <span class="dt">mesenvsimmuno =</span><span class="st">&quot;classesMesenchymal-classesImmunoreactive&quot;</span>,
                  <span class="dt">levels=</span>modelDesign)
fit_glm &lt;-<span class="st"> </span><span class="kw">glmFit</span>(d,modelDesign)
mesenvsimmuno &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit_glm , <span class="dt">contrast =</span> contrast_mesenvsimmuno)
tt_mesenvsimmuno &lt;-<span class="st"> </span><span class="kw">topTags</span>(mesenvsimmuno,<span class="dt">n=</span><span class="kw">nrow</span>(d))</code></pre></div>
<p>Examples of different designs that can be used for this dataset. Instead of simple two class design you can also compare one class to the remaining three classes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">classes &lt;-<span class="st"> </span><span class="kw">factor</span>(classDefinitions_RNASeq[,data_classes])
modelDesign &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="st"> </span><span class="dv">0</span> <span class="op">+</span><span class="st"> </span>classes)

contrast_immuno &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(
  <span class="dt">immunovsrest =</span><span class="st">&quot;classesImmunoreactive-(classesMesenchymal + </span>
<span class="st">  classesProliferative +classesDifferentiated)/3&quot;</span>,
  <span class="dt">levels=</span>modelDesign)
fit_glm &lt;-<span class="st"> </span><span class="kw">glmFit</span>(d,modelDesign)
immunovsrest &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit_glm , <span class="dt">contrast =</span> contrast_immuno)
tt_immunovsrest &lt;-<span class="st"> </span><span class="kw">topTags</span>(immunovsrest,<span class="dt">n=</span><span class="kw">nrow</span>(d))

contrast_mesen &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>( 
  <span class="dt">mesenvsrest =</span> <span class="st">&quot;classesMesenchymal-(classesImmunoreactive + </span>
<span class="st">  classesProliferative +classesDifferentiated)/3&quot;</span>,
  <span class="dt">levels=</span>modelDesign)
fit_glm &lt;-<span class="st"> </span><span class="kw">glmFit</span>(d,modelDesign)
mesenvsrest &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit_glm , <span class="dt">contrast =</span> contrast_mesen)
tt_mesenvsrest &lt;-<span class="st"> </span><span class="kw">topTags</span>(mesenvsrest,<span class="dt">n=</span><span class="kw">nrow</span>(d))

contrast_prolif &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>( 
  <span class="dt">prolifvsrest =</span> <span class="st">&quot;classesProliferative-(classesMesenchymal + </span>
<span class="st">  classesImmunoreactive +classesDifferentiated)/3&quot;</span>,
  <span class="dt">levels=</span>modelDesign)
fit_glm &lt;-<span class="st"> </span><span class="kw">glmFit</span>(d,modelDesign)
prolifvsrest &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit_glm , <span class="dt">contrast =</span> contrast_prolif)
tt_prolifvsrest &lt;-<span class="st"> </span><span class="kw">topTags</span>(prolifvsrest,<span class="dt">n=</span><span class="kw">nrow</span>(d))

contrast_diff &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>( 
  <span class="dt">diffvsrest =</span> <span class="st">&quot;classesDifferentiated-(classesMesenchymal + </span>
<span class="st">  classesImmunoreactive +classesProliferative)/3&quot;</span>,
  <span class="dt">levels=</span>modelDesign)
fit_glm &lt;-<span class="st"> </span><span class="kw">glmFit</span>(d,modelDesign)
diffvsrest &lt;-<span class="st"> </span><span class="kw">glmLRT</span>(fit_glm , <span class="dt">contrast =</span> contrast_diff)
tt_diffvsrest &lt;-<span class="st"> </span><span class="kw">topTags</span>(diffvsrest,<span class="dt">n=</span><span class="kw">nrow</span>(d))</code></pre></div>
</div>
<div id="create-gprofiler-input-list" class="section level3">
<h3><span class="header-section-number">1.2.8</span> Create g:Profiler input list</h3>
<p>8a. Create the gene list for use in g:Profiler or another thresholded enrichment tool. The list may comprise all genes that have a significant FDR-corrected p-value (code shown below), all significant and FDR-corrected up-regulated genes and all down-regulated genes separately, or some other combination of thresholds. Also see analogous step in the microarray protocol.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tt &lt;-<span class="st"> </span>tt_exact_test

<span class="co">#get the indices of scored dataset that have FDR &lt; 0.05</span>
select_genes =<span class="st"> </span><span class="kw">which</span>(tt<span class="op">$</span>table<span class="op">$</span>FDR<span class="op">&lt;</span><span class="fl">0.05</span>)

<span class="co">#output how many genes there are in the set that have FDR&lt;0.05</span>
<span class="kw">length</span>(select_genes)</code></pre></div>
<pre><code>## [1] 5846</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#gene names from the TCGA set contain gene name and entrez gene ids separated by ‘|’</span>
<span class="co"># for all subsequent enrichment analysis we need to have just one id.  Separate the names </span>
<span class="co"># into their two ids and keep the gene symbols</span>
topgenes_qvalue005 &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(<span class="kw">rownames</span>(tt<span class="op">$</span>table)[select_genes], 
                             <span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))

<span class="co">#output the top 5 entries in the list of top genes</span>
<span class="kw">head</span>(topgenes_qvalue005)</code></pre></div>
<pre><code>## [1] &quot;IGDCC3&quot;  &quot;ANTXR1&quot;  &quot;AEBP1&quot;   &quot;FBN1&quot;    &quot;ANGPTL2&quot; &quot;COL16A1&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#write results out to the file.  This is an example of a set that can be used for</span>
<span class="co"># Protocol 1</span>
<span class="kw">write.table</span>(topgenes_qvalue005, <span class="st">&quot;MesenchymalvsImmunoreactive_RNAseq_allsignificantgenes.txt&quot;</span>, 
            <span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="create-gsea-input-list" class="section level3">
<h3><span class="header-section-number">1.2.9</span> Create GSEA input list</h3>
<p>8b. Create a two-column rank (.RNK) file of all gene IDs and corresponding scores to for GSEA pre-ranked analysis. One option is to rank genes by t-statistic of differential gene expression. GSEA will look for enrichment in the set of most differentially expressed genes at the top of the list as well as those at the bottom of the list. Genes at the top of the list are more highly expressed in class A of samples (e.g., mesenchymal) while genes at the bottom are highly expressed in class B (e.g., immunoreactive). An alternative score can be computed by multiplying direction (sign) of fold change and logarithm of p-value for each gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#calculate ranks</span>
ranks_RNAseq =<span class="st"> </span><span class="kw">sign</span>(tt<span class="op">$</span>table<span class="op">$</span>logFC) <span class="op">*</span><span class="st"> </span><span class="op">-</span><span class="kw">log10</span>(tt<span class="op">$</span>table<span class="op">$</span>PValue)

<span class="co">#gene names from the TCGA set contain gene name and entrez gene ids separated by ‘|’</span>
<span class="co"># for all subsequent enrichment analysis we need to have just one id.  Separate the names </span>
<span class="co"># into their two ids.</span>
genenames &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table), <span class="cf">function</span>(data) 
  {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
geneids &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table),  
                          <span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">2</span>]})) 

<span class="co">#create ranks file</span>
ranks_RNAseq &lt;-<span class="st"> </span><span class="kw">cbind</span>(genenames, ranks_RNAseq)
<span class="kw">colnames</span>(ranks_RNAseq) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GeneName&quot;</span>,<span class="st">&quot;rank&quot;</span>)

<span class="co">#sort ranks in decreasing order</span>
ranks_RNAseq &lt;-<span class="st"> </span>ranks_RNAseq[<span class="kw">order</span>(<span class="kw">as.numeric</span>(ranks_RNAseq[,<span class="dv">2</span>]),<span class="dt">decreasing =</span> <span class="ot">TRUE</span>),]

<span class="kw">write.table</span>(ranks_RNAseq, <span class="kw">paste</span>(working_dir,
                                <span class="st">&quot;Supplementary_Table2_MesenvsImmuno_RNASeq_ranks.rnk&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
            <span class="dt">col.name =</span> <span class="ot">TRUE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="create-expression-file" class="section level3">
<h3><span class="header-section-number">1.2.10</span> Create expression file</h3>
<p>9a. Create an expression file for the enrichment map and save it to a file in the working folder. The optional expression file is similar to the expression matrix except for an additional column on the left edge of the matrix. The field often includes gene description however any text value can be added.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#fix issue with biomart not working because of url redirection</span>
<span class="kw">options</span>(<span class="dt">RCurlOptions=</span><span class="kw">list</span>(<span class="dt">followlocation=</span><span class="ot">TRUE</span>, <span class="dt">postredir=</span>2L))

normalized_expression_RNAseq &lt;-<span class="st"> </span><span class="kw">cpm</span>(d, <span class="dt">normalized.lib.size=</span><span class="ot">TRUE</span>)

<span class="co">#From the rownames parse out the gene name and the geneids</span>
genenames &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(normalized_expression_RNAseq), 
<span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
geneids &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(normalized_expression_RNAseq), 
<span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">2</span>]}))

EM_expressionFile_RNAseq &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">Name =</span> genenames, normalized_expression_RNAseq)
<span class="kw">rownames</span>(EM_expressionFile_RNAseq) &lt;-<span class="st"> </span><span class="kw">rownames</span>(normalized_expression_RNAseq)
<span class="kw">colnames</span>(EM_expressionFile_RNAseq) &lt;-<span class="st"> </span><span class="kw">substring</span>(<span class="kw">colnames</span>(EM_expressionFile_RNAseq),<span class="dv">1</span>,<span class="dv">12</span>)

<span class="co">#Add descriptions instead of geneids</span>
<span class="kw">tryCatch</span>(<span class="dt">expr =</span> { <span class="kw">library</span>(<span class="st">&quot;biomaRt&quot;</span>)}, 
         <span class="dt">error =</span> <span class="cf">function</span>(e) { 
           <span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
           <span class="kw">biocLite</span>(<span class="st">&quot;biomaRt&quot;</span>)}, 
         <span class="dt">finally =</span> <span class="kw">library</span>(<span class="st">&quot;biomaRt&quot;</span>))
mart =<span class="st"> </span><span class="kw">useMart</span>(<span class="dt">biomart=</span><span class="st">&quot;ENSEMBL_MART_ENSEMBL&quot;</span>)
mart =<span class="st"> </span><span class="kw">useDataset</span>(mart, <span class="dt">dataset=</span><span class="st">&quot;hsapiens_gene_ensembl&quot;</span> )

genes =<span class="st"> </span><span class="kw">getBM</span>(<span class="dt">attributes =</span> <span class="kw">c</span>( <span class="st">&#39;hgnc_symbol&#39;</span>, <span class="st">&#39;description&#39;</span>), <span class="dt">filters=</span><span class="st">&#39;hgnc_symbol&#39;</span>, 
              <span class="dt">values=</span>genenames, <span class="dt">mart=</span>mart);
genes<span class="op">$</span>description =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">[Source.*&quot;</span>, <span class="st">&quot;&quot;</span>, genes<span class="op">$</span>description);

EM_expressionFile_RNAseq &lt;-<span class="st"> </span><span class="kw">merge</span>(genes,EM_expressionFile_RNAseq,  <span class="dt">all.y=</span><span class="ot">TRUE</span>,<span class="dt">by.x=</span><span class="dv">1</span>, <span class="dt">by.y=</span><span class="dv">1</span>)
<span class="kw">colnames</span>(EM_expressionFile_RNAseq)[<span class="dv">1</span>] &lt;-<span class="st"> &quot;Name&quot;</span>
<span class="kw">colnames</span>(EM_expressionFile_RNAseq)[<span class="dv">2</span>] &lt;-<span class="st"> &quot;Description&quot;</span>

<span class="kw">write.table</span>(EM_expressionFile_RNAseq, 
            <span class="kw">paste</span>(working_dir,
                  <span class="st">&quot;Supplementary_Table6_TCGA_OV_RNAseq_expression.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;/&quot;</span>),
            <span class="dt">col.name=</span><span class="ot">TRUE</span>,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)

<span class="co">#write out a GSEA classes file. (optional)</span>
fileConn &lt;-<span class="st"> </span><span class="kw">file</span>(
  <span class="kw">paste</span>(working_dir,<span class="st">&quot;Supplementary_Table9_TCGA_OV_RNAseq_classes.cls&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>))
<span class="kw">writeLines</span>(<span class="kw">c</span>(<span class="kw">paste</span>(<span class="kw">length</span>(classDefinitions_RNASeq[,data_classes]), <span class="st">&quot;4 1&quot;</span>), 
             <span class="kw">paste</span>(<span class="st">&quot;# &quot;</span>, <span class="kw">unique</span>(classDefinitions_RNASeq[,data_classes])[<span class="dv">1</span>], <span class="st">&quot; &quot;</span>,
                   <span class="kw">unique</span>(classDefinitions_RNASeq[,data_classes])[<span class="dv">2</span>], <span class="st">&quot; &quot;</span>,
                   <span class="kw">unique</span>(classDefinitions_RNASeq[,data_classes])[<span class="dv">3</span>], <span class="st">&quot; &quot;</span>,
                   <span class="kw">unique</span>(classDefinitions_RNASeq[,data_classes])[<span class="dv">4</span>])), fileConn)
<span class="kw">write.table</span>(<span class="kw">t</span>(classDefinitions_RNASeq[,data_classes]), 
            <span class="kw">paste</span>(working_dir,<span class="st">&quot;Supplementary_Table9_TCGA_OV_RNAseq_classes.cls&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
            <span class="dt">col.name=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,
            <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>, <span class="dt">append=</span><span class="ot">TRUE</span>)
<span class="kw">close</span>(fileConn)</code></pre></div>
</div>
<div id="examine-results" class="section level3">
<h3><span class="header-section-number">1.2.11</span> Examine results</h3>
<p>9b. Examine gene expression data using heat maps. Heat maps can easily show the separation between sample classes, labeled by colors in the heat map header. By limiting to the most significantly differentially expressed list of genes (FDR-corrected p&lt;0.05) we can verify whether the scoring accurately separates class A from class B.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tryCatch</span>(<span class="dt">expr =</span> { <span class="kw">library</span>(<span class="st">&quot;pheatmap&quot;</span>)}, 
         <span class="dt">error =</span> <span class="cf">function</span>(e) { 
          <span class="kw">install.packages</span>(<span class="st">&quot;pheatmap&quot;</span>)}, 
         <span class="dt">finally =</span> <span class="kw">library</span>(<span class="st">&quot;pheatmap&quot;</span>))
<span class="kw">tryCatch</span>(<span class="dt">expr =</span> { <span class="kw">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>)}, 
         <span class="dt">error =</span> <span class="cf">function</span>(e) { 
          <span class="kw">install.packages</span>(<span class="st">&quot;RColorBrewer&quot;</span>)}, 
         <span class="dt">finally =</span> <span class="kw">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>))

annotation_col &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">SUBTYPE=</span><span class="kw">factor</span>(classDefinitions_RNASeq[,data_classes]))
<span class="kw">rownames</span>(annotation_col) &lt;-<span class="st"> </span>classDefinitions_RNASeq[,<span class="dv">2</span>]

ann_colors =<span class="st"> </span><span class="kw">list</span>(<span class="dt">SUBTYPE =</span> <span class="kw">c</span>(<span class="dt">Immunoreactive=</span><span class="st">&quot;blue&quot;</span>, <span class="dt">Mesenchymal=</span><span class="st">&quot;red&quot;</span>,
                              <span class="dt">Proliferative =</span> <span class="st">&quot;orange&quot;</span>,<span class="dt">Differentiated=</span><span class="st">&quot;darkgreen&quot;</span>))
col.pal &lt;-<span class="st"> </span><span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">11</span>, <span class="st">&quot;RdBu&quot;</span>))

genes_to_select &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table)[<span class="kw">which</span>(tt<span class="op">$</span>table<span class="op">$</span>FDR<span class="op">&lt;</span><span class="fl">0.05</span>)], <span class="cf">function</span>(data) 
  {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))

matrix_for_heatmap &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(EM_expressionFile_RNAseq[<span class="kw">which</span>(EM_expressionFile_RNAseq[,<span class="dv">1</span>] 
                            <span class="op">%in%</span><span class="st"> </span>genes_to_select ),<span class="dv">3</span><span class="op">:</span><span class="kw">dim</span>(EM_expressionFile_RNAseq)[<span class="dv">2</span>] ])

<span class="kw">class</span>(matrix_for_heatmap) &lt;-<span class="st"> &quot;numeric&quot;</span>
matrix_for_heatmap[matrix_for_heatmap <span class="op">==</span><span class="st"> </span><span class="dv">0</span>] &lt;-<span class="st"> </span><span class="fl">0.0000001</span>
<span class="kw">pheatmap</span>(matrix_for_heatmap, <span class="dt">color=</span>col.pal, <span class="dt">scale=</span><span class="st">&quot;row&quot;</span>, 
         <span class="dt">kmeans_k=</span><span class="ot">NA</span>, <span class="dt">show_rownames=</span><span class="ot">FALSE</span>, <span class="dt">show_colnames=</span><span class="ot">FALSE</span>, 
         <span class="dt">main=</span><span class="st">&quot;heatmap top genes(Mesen vs Immuno)&quot;</span>, <span class="dt">cluster_rows=</span><span class="ot">TRUE</span>, 
         <span class="dt">cluster_cols=</span><span class="ot">FALSE</span>, <span class="dt">clustering_distance_rows=</span><span class="st">&quot;correlation&quot;</span>,
         <span class="dt">annotation_col=</span>annotation_col, <span class="dt">annotation_colors=</span>ann_colors)</code></pre></div>
<p><img src="supplemental_protocol1_rnaseq_files/figure-html/unnamed-chunk-13-1.png" /><!-- --></p>
</div>
<div id="example-of-other-comparisons" class="section level3">
<h3><span class="header-section-number">1.2.12</span> Example of other comparisons</h3>
<p>Example of other comparison that can be done with this dataset. Rank files for each are also created</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#Immuno vs rest</span>
tt &lt;-<span class="st"> </span>tt_immunovsrest
select_genes =<span class="st"> </span><span class="kw">which</span>(tt<span class="op">$</span>table<span class="op">$</span>FDR<span class="op">&lt;</span><span class="fl">0.05</span>)
<span class="kw">length</span>(select_genes)</code></pre></div>
<pre><code>## [1] 6119</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes_qvalue005 &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table)[select_genes], 
<span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
<span class="kw">head</span>(topgenes_qvalue005)</code></pre></div>
<pre><code>## [1] &quot;AIM2&quot;   &quot;TFEC&quot;   &quot;GIMAP4&quot; &quot;CD48&quot;   &quot;CD86&quot;   &quot;HAVCR2&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(topgenes_qvalue005, 
<span class="kw">paste</span>(working_dir,<span class="st">&quot;ImmunovsRest_RNAseq_allsignificantgenes.txt&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)

ranks_RNAseq =<span class="st"> </span><span class="kw">sign</span>(tt<span class="op">$</span>table<span class="op">$</span>logFC) <span class="op">*</span><span class="st"> </span><span class="op">-</span><span class="kw">log10</span>(tt<span class="op">$</span>table<span class="op">$</span>PValue)
genenames &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table), <span class="cf">function</span>(data) 
  {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
geneids &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table),  
                          <span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">2</span>]})) 
ranks_RNAseq &lt;-<span class="st"> </span><span class="kw">cbind</span>(genenames, ranks_RNAseq)
<span class="kw">colnames</span>(ranks_RNAseq) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GeneName&quot;</span>,<span class="st">&quot;rank&quot;</span>)
<span class="kw">write.table</span>(ranks_RNAseq, <span class="kw">paste</span>(working_dir,<span class="st">&quot;ImmunovsRest_RNASeq_ranks.rnk&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.name =</span> <span class="ot">TRUE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>)


<span class="co">#Mesen vs rest data</span>
tt &lt;-<span class="st"> </span>tt_mesenvsrest
select_genes =<span class="st"> </span><span class="kw">which</span>(tt<span class="op">$</span>table<span class="op">$</span>FDR<span class="op">&lt;</span><span class="fl">0.05</span>)
<span class="kw">length</span>(select_genes)</code></pre></div>
<pre><code>## [1] 6426</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes_qvalue005 &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table)[select_genes], 
<span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
<span class="kw">head</span>(topgenes_qvalue005)</code></pre></div>
<pre><code>## [1] &quot;CRISPLD2&quot; &quot;THBS2&quot;    &quot;FBN1&quot;     &quot;COL5A1&quot;   &quot;COL3A1&quot;   &quot;AEBP1&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(topgenes_qvalue005, 
<span class="kw">paste</span>(working_dir,<span class="st">&quot;MesenvsRest_RNAseq_allsignificantgenes.txt&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)

ranks_RNAseq =<span class="st"> </span><span class="kw">sign</span>(tt<span class="op">$</span>table<span class="op">$</span>logFC) <span class="op">*</span><span class="st"> </span><span class="op">-</span><span class="kw">log10</span>(tt<span class="op">$</span>table<span class="op">$</span>PValue)
genenames &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table), <span class="cf">function</span>(data) 
  {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
geneids &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table),  
                          <span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">2</span>]})) 
ranks_RNAseq &lt;-<span class="st"> </span><span class="kw">cbind</span>(genenames, ranks_RNAseq)
<span class="kw">colnames</span>(ranks_RNAseq) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GeneName&quot;</span>,<span class="st">&quot;rank&quot;</span>)
<span class="kw">write.table</span>(ranks_RNAseq, <span class="kw">paste</span>(working_dir,<span class="st">&quot;MesenvsRest_RNASeq_ranks.rnk&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.name =</span> <span class="ot">TRUE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>)

<span class="co">#Differentiated vs rest data</span>
tt &lt;-<span class="st"> </span>tt_diffvsrest
select_genes =<span class="st"> </span><span class="kw">which</span>(tt<span class="op">$</span>table<span class="op">$</span>FDR<span class="op">&lt;</span><span class="fl">0.05</span>)
<span class="kw">length</span>(select_genes)</code></pre></div>
<pre><code>## [1] 5142</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes_qvalue005 &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table)[select_genes], 
<span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
<span class="kw">head</span>(topgenes_qvalue005)</code></pre></div>
<pre><code>## [1] &quot;KCNRG&quot;  &quot;CDHR3&quot;  &quot;SNTN&quot;   &quot;MARCKS&quot; &quot;DNAH12&quot; &quot;KIF19&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(topgenes_qvalue005, 
<span class="kw">paste</span>(working_dir,<span class="st">&quot;DiffvsRest_RNAseq_allsignificantgenes.txt&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)

ranks_RNAseq =<span class="st"> </span><span class="kw">sign</span>(tt<span class="op">$</span>table<span class="op">$</span>logFC) <span class="op">*</span><span class="st"> </span><span class="op">-</span><span class="kw">log10</span>(tt<span class="op">$</span>table<span class="op">$</span>PValue)
genenames &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table), <span class="cf">function</span>(data) 
  {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
geneids &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table),  
                          <span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">2</span>]})) 
ranks_RNAseq &lt;-<span class="st"> </span><span class="kw">cbind</span>(genenames, ranks_RNAseq)
<span class="kw">colnames</span>(ranks_RNAseq) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GeneName&quot;</span>,<span class="st">&quot;rank&quot;</span>)
<span class="kw">write.table</span>(ranks_RNAseq, <span class="kw">paste</span>(working_dir,<span class="st">&quot;DiffvsRest_RNASeq_ranks.rnk&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.name =</span> <span class="ot">TRUE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>)

<span class="co">#Proliferative vs rest data</span>
tt &lt;-<span class="st"> </span>tt_prolifvsrest
select_genes =<span class="st"> </span><span class="kw">which</span>(tt<span class="op">$</span>table<span class="op">$</span>FDR<span class="op">&lt;</span><span class="fl">0.05</span>)
<span class="kw">length</span>(select_genes)</code></pre></div>
<pre><code>## [1] 8610</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes_qvalue005 &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table)[select_genes], 
<span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
<span class="kw">head</span>(topgenes_qvalue005)</code></pre></div>
<pre><code>## [1] &quot;FGF17&quot;  &quot;CEL&quot;    &quot;CUX2&quot;   &quot;GFRA3&quot;  &quot;DPYSL5&quot; &quot;CPA1&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(topgenes_qvalue005, 
<span class="kw">paste</span>(working_dir,<span class="st">&quot;ProlifvsRest_RNAseq_allsignificantgenes.txt&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)

ranks_RNAseq =<span class="st"> </span><span class="kw">sign</span>(tt<span class="op">$</span>table<span class="op">$</span>logFC) <span class="op">*</span><span class="st"> </span><span class="op">-</span><span class="kw">log10</span>(tt<span class="op">$</span>table<span class="op">$</span>PValue)
genenames &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table), <span class="cf">function</span>(data) 
  {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">1</span>]}))
geneids &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>( <span class="kw">rownames</span>(tt<span class="op">$</span>table),  
                          <span class="cf">function</span>(data) {<span class="kw">unlist</span>(<span class="kw">strsplit</span>(data,<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|&quot;</span>))[<span class="dv">2</span>]})) 
ranks_RNAseq &lt;-<span class="st"> </span><span class="kw">cbind</span>(genenames, ranks_RNAseq)
<span class="kw">colnames</span>(ranks_RNAseq) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GeneName&quot;</span>,<span class="st">&quot;rank&quot;</span>)
<span class="kw">write.table</span>(ranks_RNAseq, <span class="kw">paste</span>(working_dir,<span class="st">&quot;ProlifvsRest_RNASeq_ranks.rnk&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), 
<span class="dt">col.name =</span> <span class="ot">TRUE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-TCGA">
<p>1. International Cancer Genome, C. <em>et al.</em> International network of cancer genome projects. <em>Nature</em> <strong>464,</strong> 993–8 (2010).</p>
</div>
<div id="ref-GDC">
<p>2. Grossman, R. L. <em>et al.</em> Toward a shared vision for cancer genomic data. <em>N Engl J Med</em> <strong>375,</strong> 1109–12 (2016).</p>
</div>
<div id="ref-TCGABiolinks">
<p>3. Colaprico, A. <em>et al.</em> TCGAbiolinks: An r/bioconductor package for integrative analysis of tcga data. <em>Nucleic Acids Res</em> <strong>44,</strong> e71 (2016).</p>
</div>
<div id="ref-MapSplice">
<p>4. Wang, K. <em>et al.</em> MapSplice: Accurate mapping of rna-seq reads for splice junction discovery. <em>Nucleic Acids Res</em> <strong>38,</strong> e178 (2010).</p>
</div>
<div id="ref-RSEM">
<p>5. Li, B. &amp; Dewey, C. N. RSEM: Accurate transcript quantification from rna-seq data with or without a reference genome. <em>BMC Bioinformatics</em> <strong>12,</strong> 323 (2011).</p>
</div>
<div id="ref-OV">
<p>6. Verhaak, R. G. <em>et al.</em> Prognostically relevant gene signatures of high-grade serous ovarian carcinoma. <em>J Clin Invest</em> <strong>123,</strong> 517–25 (2013).</p>
</div>
<div id="ref-edgeR">
<p>7. Robinson, M. D., McCarthy, D. J. &amp; Smyth, G. K. EdgeR: A bioconductor package for differential expression analysis of digital gene expression data. <em>Bioinformatics</em> <strong>26,</strong> 139–40 (2010).</p>
</div>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
