<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Ruth Isserlin" />

<meta name="date" content="2017-12-12" />

<title>Supplementary Protocol 2 – create a gene list by analyzing gene expression data from Affymetrix microarrays with Limma</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #0000ff; } /* Keyword */
code > span.ch { color: #008080; } /* Char */
code > span.st { color: #008080; } /* String */
code > span.co { color: #008000; } /* Comment */
code > span.ot { color: #ff4000; } /* Other */
code > span.al { color: #ff0000; } /* Alert */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #008000; font-weight: bold; } /* Warning */
code > span.cn { } /* Constant */
code > span.sc { color: #008080; } /* SpecialChar */
code > span.vs { color: #008080; } /* VerbatimString */
code > span.ss { color: #008080; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #0000ff; } /* ControlFlow */
code > span.op { } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ff4000; } /* Preprocessor */
code > span.do { color: #008000; } /* Documentation */
code > span.an { color: #008000; } /* Annotation */
code > span.cv { color: #008000; } /* CommentVar */
code > span.at { } /* Attribute */
code > span.in { color: #008000; } /* Information */
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 64px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 69px;
  margin-top: -69px;
}

.section h2 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h3 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h4 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h5 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h6 {
  padding-top: 69px;
  margin-top: -69px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Enrichment Map Protocol</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fa fa-home"></span>
     
    Home
  </a>
</li>
<li>
  <a href="https://www.biorxiv.org/content/early/2017/12/12/232835">
    <span class="fa fa-newspaper-o"></span>
     
    bioRxiv paper
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    <span class="fa fa-gear"></span>
     
    More
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Supplementary Protocols</li>
    <li>
      <a href="Download_TCGA_data.html">Download TCGA Data</a>
    </li>
    <li>
      <a href="supplemental_protocol1_rnaseq.html">Process RNASeq data</a>
    </li>
    <li>
      <a href="supplemental_protocol2_microarray.html">Process Microarray data</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Protocol 2</li>
    <li>
      <a href="Protocol2_createEM.html">Create Enrichment Map</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Supplementary Protocol 2 – create a gene list by analyzing gene expression data from Affymetrix microarrays with Limma</h1>
<h4 class="author"><em>Ruth Isserlin</em></h4>
<h4 class="date"><em>2017-12-12</em></h4>

</div>


<div id="process-microarray-data" class="section level1">
<h1><span class="header-section-number">1</span> Process Microarray data</h1>
<p>This protocol demonstrates the extraction of gene lists for pathway enrichment analysis using RMA-normalized gene expression data from Affymetrix microarrays for downstream pathway enrichment analysis with g:Profiler, GSEA and other similar tools. g:Profiler requires a ranked list of differentially expressed genes that are filtered according to a significance cut-off. GSEA requires a two-column tab-separated RNK file with a ranked list of all genes in the genome. In the RNK file, the first column specifies the gene name and the second column specifies a numeric score representing the level of differential expression. For both methods, the first step involves calculating a statistic for each gene that represents the difference in its expression levels between the two groups. This step is performed using the limma R package.</p>
<div id="load-required-packages" class="section level2">
<h2><span class="header-section-number">1.1</span> Load required packages</h2>
<ol start="9" style="list-style-type: decimal">
<li>Load required Bioconductor packages into R</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#check to see if the library has already been installed.  If it hasn&#39;t then install it. </span>
<span class="kw">tryCatch</span>(<span class="dt">expr =</span> { <span class="kw">library</span>(<span class="st">&quot;Biobase&quot;</span>)}, 
         <span class="dt">error =</span> function(e) { 
           <span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
           <span class="kw">biocLite</span>(<span class="st">&quot;Biobase&quot;</span>)}, 
         <span class="dt">finally =</span> <span class="kw">library</span>(<span class="st">&quot;Biobase&quot;</span>))</code></pre></div>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, cbind, colMeans,
##     colnames, colSums, do.call, duplicated, eval, evalq, Filter,
##     Find, get, grep, grepl, intersect, is.unsorted, lapply,
##     lengths, Map, mapply, match, mget, order, paste, pmax,
##     pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce,
##     rowMeans, rownames, rowSums, sapply, setdiff, sort, table,
##     tapply, union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">tryCatch</span>(<span class="dt">expr =</span> { <span class="kw">library</span>(<span class="st">&quot;limma&quot;</span>)}, 
         <span class="dt">error =</span> function(e) { 
           <span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
           <span class="kw">biocLite</span>(<span class="st">&quot;limma&quot;</span>)}, 
         <span class="dt">finally =</span> <span class="kw">library</span>(<span class="st">&quot;limma&quot;</span>))</code></pre></div>
<pre><code>## Warning: package &#39;limma&#39; was built under R version 3.4.1</code></pre>
<pre><code>## 
## Attaching package: &#39;limma&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     plotMA</code></pre>
<ol start="10" style="list-style-type: decimal">
<li>Set the working directory to the location of where the Supplemental Tables 1-4 are stored. The function getwd() shows the working directory and dir() shows its files.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">working_dir &lt;-<span class="st"> &quot;./data&quot;</span></code></pre></div>
</div>
<div id="load-expression-data" class="section level2">
<h2><span class="header-section-number">1.2</span> Load Expression Data</h2>
<ol start="11" style="list-style-type: decimal">
<li>Load expression data into R. Minimally the expression set requires a gene name for each row and typically at least 6 expression values (3 values in each compared class). Our dataset consists of 216 patients with 107 immunoreactive and 109 mesenchymal samples. After loading, use the command head(expressionMatrix) to verify the loaded matrix.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">expressionMatrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">read.table</span>(
  <span class="kw">paste</span>(working_dir, <span class="st">&quot;Supplementary_Table10_TCGA_Microarray_rmanormalized.txt&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;</span><span class="ch">\&quot;</span><span class="st">&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>))</code></pre></div>
</div>
<div id="load-subtype-information" class="section level2">
<h2><span class="header-section-number">1.3</span> Load subtype information</h2>
<ol start="12" style="list-style-type: decimal">
<li>Load subtype classification of samples. To calculate differential expression, we need to define at least 2 classes of samples. A common experimental design involves cases and controls but any two classes can be used. The current dataset is divided into mesenchymal and immunoreactive classes (Supplementary Table 3, third column). After loading the matrix, check that the column names of the expression matrix and class definitions are equal.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">classDefinitions &lt;-<span class="st"> </span><span class="kw">read.table</span>( <span class="kw">paste</span>(working_dir,<span class="st">&quot;Supplementary_Table11_Microarray_classdefinitions.txt&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;/&quot;</span>), <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">quote=</span><span class="st">&quot;</span><span class="ch">\&quot;</span><span class="st">&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
<span class="kw">identical</span>(<span class="kw">colnames</span>(expressionMatrix), classDefinitions$patient)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="calculate-differential-expression" class="section level2">
<h2><span class="header-section-number">1.4</span> Calculate Differential expression</h2>
<ol start="13" style="list-style-type: decimal">
<li>Format data and class definitions for limma. The expression data needs to be converted to an object of type ExpressionSet. The ExpressionSet needs to include a data matrix where rows are genes, columns are samples and each cell contains an expression value. Classes need to be defined as factors.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">minimalSet &lt;-<span class="st"> </span><span class="kw">ExpressionSet</span>(<span class="dt">assayData=</span>expressionMatrix)
classes &lt;-<span class="st"> </span><span class="kw">factor</span>(classDefinitions[,<span class="st">&quot;SUBTYPE&quot;</span>])</code></pre></div>
<ol start="14" style="list-style-type: decimal">
<li>Create a model matrix with the defined classes.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">modelDesign &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(~<span class="st"> </span><span class="dv">0</span> +<span class="st"> </span>classes)</code></pre></div>
<ol start="15" style="list-style-type: decimal">
<li>Fit the model to the expression matrix.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">lmFit</span>(minimalSet, modelDesign)</code></pre></div>
<ol start="16" style="list-style-type: decimal">
<li>Create the contrast matrix - By specifying Mesenchymal first and Immunoreactive second, positive logFC and t-values refer to higher expression levels (up-regulation) in the Mesenchenchymal versus Immunoreactive samples</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">contrastnm &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;classesMesenchymal-classesImmunoreactive&quot;</span>) 
contrast.matrix &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(<span class="dt">original =</span><span class="st">&quot;classesMesenchymal-classesImmunoreactive&quot;</span>, 
                                 <span class="dt">mesenvsrest =</span> <span class="st">&quot;classesMesenchymal-(classesImmunoreactive + classesProliferative +classesDifferentiated)/3&quot;</span>,
                                 <span class="dt">immunovsrest =</span> <span class="st">&quot;classesImmunoreactive-(classesMesenchymal + classesProliferative +classesDifferentiated)/3&quot;</span>,
                                 <span class="dt">prolifvsrest =</span> <span class="st">&quot;classesProliferative-(classesMesenchymal + classesImmunoreactive +classesDifferentiated)/3&quot;</span>, 
                                 <span class="dt">diffvsrest =</span> <span class="st">&quot;classesDifferentiated-(classesMesenchymal + classesImmunoreactive +classesProliferative)/3&quot;</span>,<span class="dt">levels=</span>modelDesign)
<span class="co">#contrast.matrix &lt;- makeContrasts(mesenvsrest = &quot;classesMesenchymal-(classesImmunoreactive + classesProliferative +classesDifferentiated)/3&quot;,levels=modelDesign)</span></code></pre></div>
<ol start="17" style="list-style-type: decimal">
<li>Model contrasts of gene expression. The following command models gene expression differences of each gene between the two groups of samples using linear regression and computes coefficients and standard errors.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit1 &lt;-<span class="st"> </span><span class="kw">contrasts.fit</span>(fit, contrast.matrix)</code></pre></div>
<ol start="18" style="list-style-type: decimal">
<li>Compute differential expression statistics. Given a fitted linear regression model, the command generates a table containing the log fold change, average expression, t statistic, p-value, adjusted p-value and B statistic for each entity in the expression matrix using empirical Bayes statistics. The B-statistic represents the log-odds that the gene is differentially expressed but it is based on a prior assumption of how many genes are differentially expressed in the dataset. Because of its reliance on this prior assumption, the adjusted p-value is preferentially used as an indicator of significant differential expression.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit2 &lt;-<span class="st"> </span><span class="kw">eBayes</span>(fit1)</code></pre></div>
<ol start="19" style="list-style-type: decimal">
<li>Generate a table with differentially expressed genes and adjust for multiple hypothesis testing using Benjamini-Hochberg False Discovery Rate. The table contains all genes ranked by p-value and shown with log fold change, average expression, t-statistic, p-value, adjusted p-value and B-statistic.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topfit &lt;-<span class="st"> </span><span class="kw">topTable</span>(fit2, <span class="dt">coef=</span><span class="st">&quot;immunovsrest&quot;</span>,<span class="dt">number=</span><span class="kw">nrow</span>(expressionMatrix), <span class="dt">adjust=</span><span class="st">&quot;BH&quot;</span>)</code></pre></div>
</div>
<div id="create-gprofiler-input-list" class="section level2">
<h2><span class="header-section-number">1.5</span> Create g:Profiler input list</h2>
<p>20a. Create the gene list for use in g:Profiler or another thresholded enrichment tool. The list may comprise all genes that have a significant FDR-corrected p-value, all up-regulated genes with a significant FDR p-value, all down-regulated genes with a significant FDR p-value, or some other combination of thresholds. * To get all significant genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">which</span>(topfit$adj.P.Val&lt;<span class="fl">0.05</span>))</code></pre></div>
<pre><code>## [1] 3988</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes_qvalue005 &lt;-<span class="st"> </span><span class="kw">rownames</span>(topfit)[<span class="kw">which</span>(topfit$adj.P.Val&lt;<span class="fl">0.05</span>)]
<span class="kw">head</span>(topgenes_qvalue005)</code></pre></div>
<pre><code>## [1] &quot;TFEC&quot;  &quot;CD86&quot;  &quot;CD48&quot;  &quot;IL2RG&quot; &quot;LCP2&quot;  &quot;GPR65&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(topgenes_qvalue005, 
<span class="st">&quot;ImmunovsRest_allsignificantgenes.txt&quot;</span>, 
<span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="examine-results" class="section level2">
<h2><span class="header-section-number">1.6</span> Examine results</h2>
<ul>
<li>Significantly up-regulated genes in mesenchymal samples have positive logFC and t-values.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">which</span>(topfit$adj.P.Val&lt;<span class="fl">0.05</span> &amp;<span class="st"> </span>topfit$t &gt;<span class="dv">0</span>))</code></pre></div>
<pre><code>## [1] 2235</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes_qvalue005_mesenchymal &lt;-
<span class="st">    </span><span class="kw">rownames</span>(topfit)[<span class="kw">which</span>(topfit$adj.P.Val&lt;<span class="fl">0.05</span> &amp;<span class="st"> </span>topfit$t &gt;<span class="dv">0</span>)]
<span class="kw">head</span>(topgenes_qvalue005_mesenchymal)</code></pre></div>
<pre><code>## [1] &quot;TFEC&quot;  &quot;CD86&quot;  &quot;CD48&quot;  &quot;IL2RG&quot; &quot;LCP2&quot;  &quot;GPR65&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(topgenes_qvalue005_mesenchymal, <span class="st">&quot;ImmunovsRest_immuno_significantgenes.txt&quot;</span>, <span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)</code></pre></div>
<ul>
<li>Significantly up-regulated genes in immunoreactive samples have negative logFC and t-values.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(<span class="kw">which</span>(topfit$adj.P.Val&lt;<span class="fl">0.05</span> &amp;<span class="st"> </span>topfit$t &lt;<span class="dv">0</span>))</code></pre></div>
<pre><code>## [1] 1753</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">topgenes_qvalue005_immunoreactive&lt;-<span class="st"> </span><span class="kw">rownames</span>(topfit)[<span class="kw">which</span>(topfit$adj.P.Val&lt;<span class="fl">0.05</span> &amp;<span class="st"> </span>topfit$t &lt;<span class="dv">0</span>)]
<span class="kw">head</span>(topgenes_qvalue005_immunoreactive)</code></pre></div>
<pre><code>## [1] &quot;GPC1&quot;   &quot;SH3BP4&quot; &quot;TCF7L1&quot; &quot;LAMB1&quot;  &quot;MYH10&quot;  &quot;TTC3&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">write.table</span>(topgenes_qvalue005_immunoreactive, 
<span class="st">&quot;ImmunovsRest_rest_significantgenes.txt&quot;</span>, <span class="dt">col.names=</span><span class="ot">FALSE</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="create-gsea-input-list" class="section level2">
<h2><span class="header-section-number">1.7</span> Create GSEA input list</h2>
<p>20b. Create a rank file for GSEA. To run GSEA in pre-ranked mode, you need a two column RNK file with gene/protein/probe name (column 1) and the associated score (column 2). The first column should contain the same type of gene IDs used in the pathway gene-set (GMT) file. GSEA looks for enrichment in the top and bottom parts of the list, ranking the file using the t-statistic. The t-statistic indicates the strength of differential expression and is used in the p-value calculation. Other scores indicating the strength of differential expression may be used as well. GSEA ranks the most up-regulated genes at the top of the list and the most down-regulated at the bottom of the list. Genes at the top of the list are more highly expressed in class A compared to class B, while genes at the bottom of the list are higher in class B. In this workflow, a positive t-value means a higher expression of a gene in the Mesenchymal samples compared to the Immunoreactive samples (variable constrastnm). The following commands create a data frame with gene IDs and t-statistics, remove lines with missing gene IDs, and store the result as a RNK file. An additional step is usually required in analysis of Affymetrix microarray data as genes are represented with multiple probesets. The most significant probeset or average probeset score may be considered for every gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ranks &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">geneID=</span><span class="kw">rownames</span>(topfit),<span class="dt">t_stat=</span>topfit[,<span class="st">&quot;t&quot;</span>], <span class="dt">stringsAsFactors=</span>F)
ranks &lt;-<span class="st"> </span>ranks[<span class="kw">which</span>(ranks[,<span class="st">&quot;geneID&quot;</span>] !=<span class="st"> &quot;&quot;</span>),]
<span class="kw">write.table</span>(ranks,<span class="st">&quot;ImmunovsRest_limma_ranks.rnk&quot;</span>,
<span class="dt">col.name=</span><span class="ot">TRUE</span>,<span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">row.names=</span><span class="ot">FALSE</span>, <span class="dt">quote=</span><span class="ot">FALSE</span>)

<span class="kw">head</span>(ranks)</code></pre></div>
<pre><code>##   geneID   t_stat
## 1   TFEC 20.75773
## 2   CD86 17.36037
## 3   CD48 17.30732
## 4  IL2RG 16.87160
## 5   LCP2 16.73821
## 6  GPR65 16.65257</code></pre>
</div>
<div id="create-expression-file" class="section level2">
<h2><span class="header-section-number">1.8</span> Create expression file</h2>
<ol start="21" style="list-style-type: decimal">
<li>Create an expression file for the enrichment map and save files to the home folder of the analysis. The expression file contains the gene IDs as the first column gene description as the second column and the expression values for each sample as the additional columns. Gene IDs should correspond to the first column of the rank file. The text files will be saved on your computer in the directory specified at the beginning of the script using setwd(). The .rnk, .cls and .txt are all tab delimited files that can be viewed in spreadsheet or in a text editor.</li>
</ol>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
